### ISSUE TITLE
unhelpful error message invoking fn foo<T>() when T is not used within signature

### ISSUE THREAD

**Issue by whitequark** [rust-lang/rust#28523]
Test case:
```rust
fn foo<T>() {}
fn main() { foo() }
```
Error:
```
t.rs:2:13: 2:16 error: unable to infer enough type information about `_`; type annotations or generic parameter binding required [E0282]
t.rs:2 fn main() { foo() }
                   ^~~
```
It should, at least, mention `T` by name, and I think some rewording would also be beneficial.

**Comment by ebfull** [comment]
It doesn't matter that `T` isn't used. This also triggers the error:

```rust
fn foo<T>() {
    println!("The size of T is {}", std::mem::size_of::<T>());
}
fn main() { foo() }
```

At the call site it has to monomorphize `foo<T>()` but it cannot infer for `T`. You need to provide `foo`, at the very least, with a type parameter, like so:

```rust
fn foo<T>() {}
fn main() { foo::<usize>() }
```

**Comment by whitequark** [comment]
Yes, I understand why this is an error. The point is that the error _message_ is completely unhelpful. I spent at least an hour figuring this out.

**Comment by ebfull** [comment]
I agree the error message can be improved. I was talking about your title.

#25633 is another example of this.

**Comment by nagisa** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/28153
